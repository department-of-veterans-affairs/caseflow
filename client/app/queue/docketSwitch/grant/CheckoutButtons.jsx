import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { css } from 'glamor';
import cx from 'classnames';
import DecisionViewFooter from 'app/queue/components/DecisionViewFooter';
export const CheckoutButtons = ({
  onCancel,
  onBack,
  onSubmit,
  submitText,
  disabled = false,
  loading = false,
}) => {
  const cancelBtn = {
    classNames: [cx('cf-btn-link', { 'cf-right-side': !onBack })],
    callback: onCancel,
    name: 'cancel-button',
    displayText: 'Cancel',
    willNeverBeLoading: true,
  };
  const submitBtn = {
    classNames: ['cf-right-side'],
    callback: onSubmit,
    name: 'next-button',
    disabled,
    displayText: submitText ?? 'Continue',
    loading,
    styling: css({ marginLeft: '1rem' }),
  };
  const backBtn = {
    classNames: ['cf-right-side', 'cf-prev-step', 'usa-button-secondary'],
    callback: onBack,
    name: 'back-button',
    displayText: 'Back',
    willNeverBeLoading: true,
  };
  // Button layout from QueueFlowPage
  const buttons = useMemo(() => [
    submitBtn,
    cancelBtn,
    // Only display "Back" button if applicable
    ...(onBack && !loading ? [backBtn] : []),
  ]);

  return <DecisionViewFooter buttons={buttons} />;
};
CheckoutButtons.propTypes = {
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  onBack: PropTypes.func,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
  submitText: PropTypes.string,
};
export default CheckoutButtons;
