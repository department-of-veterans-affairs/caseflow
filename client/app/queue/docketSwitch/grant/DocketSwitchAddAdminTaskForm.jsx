import React from 'react';
import PropTypes from 'prop-types';

import { Controller, useFormContext } from 'react-hook-form';
import SearchableDropdown from 'app/components/SearchableDropdown';

import colocatedAdminActions from 'constants/CO_LOCATED_ADMIN_ACTIONS';
import StringUtil from 'app/util/StringUtil';
import TextareaField from 'app/components/TextareaField';
import {
  ADD_COLOCATED_TASK_INSTRUCTIONS_LABEL,
  FORM_ERROR_FIELD_REQUIRED,
} from 'app/../COPY';
import { css } from 'glamor';
import Button from 'app/components/Button';

const fieldStyles = css({
  marginTop: '4rem',
});

const actionOptions = Object.entries(colocatedAdminActions).map(
  ([value, label]) => ({ label, value })
);

const { capitalizeFirst, snakeCaseToCamelCase } = StringUtil;

export const DocketSwitchAddAdminTaskForm = ({ baseName, item, onRemove }) => {
  const { control, errors, register } = useFormContext();

  const handleRemove = () => onRemove();

  // We need to submit an actual task name, so reformatting is necessary
  const formatTaskName = (taskStr) =>
    `${capitalizeFirst(snakeCaseToCamelCase(taskStr))}ColocatedTask`;

  return (
    <>
      <div className={fieldStyles}>
        <Controller
          name={`${baseName}.type`}
          control={control}
          defaultValue={item.type}
          render={({ onChange, ...rest }) => (
            <SearchableDropdown
              {...rest}
              label="Select the type of task you'd like to open:"
              options={actionOptions}
              onChange={(valObj) => onChange(formatTaskName(valObj?.value))}
            />
          )}
        />
      </div>

      <div className={fieldStyles}>
        <TextareaField
          errorMessage={
            errors?.[baseName]?.instructions ? FORM_ERROR_FIELD_REQUIRED : null
          }
          name={`${baseName}.instructions`}
          defaultValue={item.instructions}
          label={ADD_COLOCATED_TASK_INSTRUCTIONS_LABEL}
          inputRef={register()}
        />
      </div>
      <div>
        <Button linkStyling onClick={handleRemove}>
          Remove this action
        </Button>
      </div>
    </>
  );
};

DocketSwitchAddAdminTaskForm.propTypes = {
  baseName: PropTypes.string.isRequired,
  item: PropTypes.shape({
    type: PropTypes.shape({ value: PropTypes.string, label: PropTypes.string }),
    instructions: PropTypes.string,
  }),
  onRemove: PropTypes.func.isRequired,
};
