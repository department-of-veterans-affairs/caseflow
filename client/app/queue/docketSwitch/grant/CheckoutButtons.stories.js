import React from 'react';

import { CheckoutButtons } from './CheckoutButtons';

export default {
  title: 'Queue/Docket Switch/CheckoutButtons',
  component: CheckoutButtons,
  decorators: [],
  parameters: {},
  args: {},
  argTypes: {
    onBack: { action: 'back' },
    onCancel: { action: 'cancel' },
    onSubmit: { action: 'submit' },
  },
};

const Template = (args) => <CheckoutButtons {...args} />;

export const Basic = Template.bind({});

export const NoBack = Template.bind({});
NoBack.args = {
  // eslint-disable-next-line no-undefined
  onBack: undefined,
};

Basic.parameters = {
  docs: {
    storyDescription: 'Used in attorney checkout flow',
  },
};
