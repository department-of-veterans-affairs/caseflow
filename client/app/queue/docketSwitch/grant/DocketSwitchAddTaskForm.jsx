import React, { useMemo, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useForm, Controller } from 'react-hook-form';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import { CheckoutButtons } from './CheckoutButtons';
import {
  DOCKET_SWITCH_GRANTED_ADD_TASK_LABEL,
  DOCKET_SWITCH_GRANTED_ADD_TASK_INSTRUCTIONS,
  DOCKET_SWITCH_GRANTED_ADD_TASK_TEXT,
  DOCKET_SWITCH_GRANTED_ADD_TASK_BUTTON
} from 'app/../COPY';
import { sprintf } from 'sprintf-js';
import { yupResolver } from '@hookform/resolvers';
import * as yup from 'yup';
import { css } from 'glamor';
import ReactMarkdown from 'react-markdown';
import CheckboxGroup from 'app/components/CheckboxGroup';
import Button from '../../../components/Button';
import StringUtil from '../../../util/StringUtil';
import DocketSwitchAddTaskModal from './DocketSwitchAddTaskModal';

const schema = yup.object().shape({
  taskIds: yup.array(yup.string())
});

export const DocketSwitchAddTaskForm = ({
  onSubmit,
  onCancel,
  docketFrom,
  docketTo,
  taskListing = [],
  onBack
}) => {
  const { handleSubmit, control, formState, setValue } = useForm({
    resolver: yupResolver(schema),
    mode: 'onChange',
    defaultValues: {
      taskIds: taskListing.map((task) => task.id),
    },
  });

  const [tasks, setTasks] = useState({});
  const [activeTaskId, setActiveTaskId] = useState(null);

  const sectionStyle = css({ marginBottom: '24px' });

  const taskOptions = useMemo(() => {
    return taskListing && taskListing.map((task) =>
      ({ label: task.label,
        id: task.taskId.toString() }));
  }, [taskListing]);

  const selectAllTasks = () => {
    const newValues = {};

    taskListing.forEach((item) => (newValues[item.taskId] = true));
    setTasks(newValues);

    setValue('taskIds', taskListing.map((task) => task.taskId));
  };

  const activeTaskLabel = useMemo(() => {
    return activeTaskId ? taskListing.find(
      (task) => String(task.taskId) === String(activeTaskId)
    )?.['label'] : null;
  }, [activeTaskId]);

  // populate all of our checkboxes on initial render
  useEffect(() => selectAllTasks(), []);

  const updateTaskSelections = (targetTaskId = null) => {
    const updatedTaskId = activeTaskId || targetTaskId;
    const newSelections = {
      ...tasks,
      [updatedTaskId]: !tasks[updatedTaskId],
    };

    setTasks(newSelections);
    setActiveTaskId(null);
    setValue(
      'taskIds',
      Object.keys(newSelections).filter((key) => newSelections[key])
    );
  };

  const handleTaskChange = (evt) => {
    const targetTaskId = evt.target.id.toString();

    setActiveTaskId(targetTaskId);

    if (!tasks[targetTaskId] === true) {
      updateTaskSelections(targetTaskId);
    }
  };

  const handleCancel = () => {
    setActiveTaskId(null);
  };

  const title = sprintf(
    DOCKET_SWITCH_GRANTED_ADD_TASK_INSTRUCTIONS,
    StringUtil.snakeCaseToCapitalized(docketFrom),
    StringUtil.snakeCaseToCapitalized(docketTo)
  );

  return (
    <form
      className="docket-switch-granted-add"
      onSubmit={handleSubmit(onSubmit)}
      aria-label="Grant Docket Switch Add Task"
    >
      <AppSegment filledBackground>
        <h1>{DOCKET_SWITCH_GRANTED_ADD_TASK_LABEL}</h1>
        <div {...sectionStyle}>
          <ReactMarkdown
            source={title}
          />
        </div>
        <div><ReactMarkdown source={DOCKET_SWITCH_GRANTED_ADD_TASK_TEXT} /></div>
        <Controller
          name="taskIds"
          control={control}
          render={({ name, onChange: onCheckChange }) => {
            return (
              <CheckboxGroup
                name={name}
                label="Please unselect any tasks you would like to remove:"
                strongLabel
                options={taskOptions}
                onChange={(event) => onCheckChange(handleTaskChange(event))}
                styling={css({ marginBottom: '0' })}
                values={tasks}
              />
            );
          }}
        />

        { activeTaskId && (
          <DocketSwitchAddTaskModal
            onCancel={handleCancel}
            taskLabel= {activeTaskLabel}
            onConfirm={updateTaskSelections}
          />
        )
        }

        <React.Fragment>
          <h3 {...css({ marginBottom: '0' })}>
            <br />
            <strong>Would you like to add any additional tasks?</strong>
            <br />
          </h3>
          <Button
            willNeverBeLoading
            dangerStyling
            styling={css({ marginTop: '1rem' })}
            name={DOCKET_SWITCH_GRANTED_ADD_TASK_BUTTON}
          />
        </React.Fragment>
      </AppSegment>
      <div className="controls cf-app-segment">
        <CheckoutButtons
          disabled={!formState.isValid}
          onCancel={onCancel}
          onBack={onBack}
          onSubmit={handleSubmit(onSubmit)}
        />
      </div>
    </form>
  );
};
DocketSwitchAddTaskForm.propTypes = {
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
  docketFrom: PropTypes.string,
  docketTo: PropTypes.string,
  taskListing: PropTypes.array,
  onBack: PropTypes.func
};
