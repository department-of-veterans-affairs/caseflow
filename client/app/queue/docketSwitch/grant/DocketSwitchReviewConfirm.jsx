import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { sprintf } from 'sprintf-js';
import { format } from 'date-fns';
import { css } from 'glamor';

import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';

import StringUtil from 'app/util/StringUtil';
import {
  DOCKET_SWITCH_GRANTED_CONFIRM_TITLE,
  DOCKET_SWITCH_GRANTED_CONFIRM_DESCRIPTION_A,
  DOCKET_SWITCH_GRANTED_CONFIRM_DESCRIPTION_B,
} from 'app/../COPY';
import CheckoutButtons from './CheckoutButtons';

const styles = {
  mainTable: css({
    '& > tbody > tr > td': {
      verticalAlign: 'top',
      ':first-child': {
        fontWeight: 'bold',
      },
    },
    '& table': {
      margin: 0,
      '& tr:first-of-type td': {
        borderTop: 'none',
      },
      '& tr:last-of-type td': {
        borderBottom: 'none',
      },
    },
  }),
};

export const DocketSwitchReviewConfirm = ({
  claimantName,
  docketFrom,
  docketTo,
  onCancel,
  onBack,
  onSubmit,
  loading = false,
  originalReceiptDate,
  docketSwitchReceiptDate,
  issuesSwitched,
  issuesRemaining,
  tasksKept,
  tasksAdded = [],
  veteranName,
}) => {
  const description1 = useMemo(
    () =>
      sprintf(
        DOCKET_SWITCH_GRANTED_CONFIRM_DESCRIPTION_A,
        StringUtil.snakeCaseToCapitalized(docketFrom),
        StringUtil.snakeCaseToCapitalized(docketTo),
        StringUtil.snakeCaseToCapitalized(docketFrom),
        StringUtil.snakeCaseToCapitalized(docketTo)
      ),
    [docketFrom, docketTo]
  );

  const noTasksShown = () => {
    if (![...tasksKept, ...tasksAdded]?.length) {
      return 'None';
    }
  };

  return (
    <>
      <AppSegment filledBackground>
        <h1>{DOCKET_SWITCH_GRANTED_CONFIRM_TITLE}</h1>
        <p>
          <strong>{description1}</strong>
        </p>
        <p>{DOCKET_SWITCH_GRANTED_CONFIRM_DESCRIPTION_B}</p>

        <table className={`usa-table-borderless ${styles.mainTable}`}>
          <tbody>
            <tr>
              <td>Veteran</td>
              <td>{veteranName}</td>
            </tr>
            {claimantName && (
              <tr>
                <td>Claimant</td>
                <td>{claimantName}</td>
              </tr>
            )}
            <tr>
              <td>VA Form - 10182 Receipt Date</td>
              <td>{format(originalReceiptDate, 'M/d/y')}</td>
            </tr>
            <tr>
              <td>Docket Switch Receipt Date</td>
              <td>{format(docketSwitchReceiptDate, 'M/d/y')}</td>
            </tr>
            <tr>
              <td>New Review option</td>
              <td>{StringUtil.snakeCaseToCapitalized(docketTo)}</td>
            </tr>
            <tr>
              <td>Issues switched to new docket</td>
              <td>
                <table className="usa-table-borderless">
                  <tbody>
                    {issuesSwitched.map((issue, idx) => (
                      <tr key={issue.id}>
                        <td>
                          <div>{`${idx + 1}. ${issue.description}`}</div>
                          <div>
                            Decision date:{' '}
                            {format(Date.parse(issue.decision_date), 'M/d/Y')}
                          </div>
                        </td>
                      </tr>
                    ))}
                  </tbody>
                </table>
              </td>
            </tr>
            {Boolean(issuesRemaining?.length) && (
              <tr>
                <td>Issues on original docket</td>
                <td>
                  <table className="usa-table-borderless">
                    <tbody>
                      {issuesRemaining.map((issue, idx) => (
                        <tr key={issue.id}>
                          <td>
                            <div>{`${idx + 1}. ${issue.description}`}</div>
                            <div>
                              Decision date:{' '}
                              {format(Date.parse(issue.decision_date), 'M/d/Y')}
                            </div>
                          </td>
                        </tr>
                      ))}
                    </tbody>
                  </table>
                </td>
              </tr>
            )}
            <tr>
              <td>Transferred/Added tasks</td>
              <td>
                <table className="usa-table-borderless">
                  <tbody>
                    <tr>
                      <td>
                        <div>
                          <em>{noTasksShown()}</em>
                        </div>
                      </td>
                    </tr>
                    {[...tasksKept, ...tasksAdded].map((task, idx) => (
                      <tr key={task.id || task.name}>
                        <td>
                          <div>{`${idx + 1}. ${task.label}`}</div>
                          <div>
                            <em>{task.instructions}</em>
                          </div>
                        </td>
                      </tr>
                    ))}
                  </tbody>
                </table>
              </td>
            </tr>
          </tbody>
        </table>
      </AppSegment>
      <div className="controls cf-app-segment">
        <CheckoutButtons
          onCancel={onCancel}
          onBack={onBack}
          onSubmit={onSubmit}
          loading={loading}
          submitText="Confirm docket switch"
        />
      </div>
    </>
  );
};
DocketSwitchReviewConfirm.propTypes = {
  veteranName: PropTypes.string,
  claimantName: PropTypes.string,
  docketFrom: PropTypes.string,
  docketTo: PropTypes.string,
  onBack: PropTypes.func,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
  loading: PropTypes.bool,
  originalReceiptDate: PropTypes.instanceOf(Date),
  docketSwitchReceiptDate: PropTypes.instanceOf(Date),
  issuesSwitched: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      date: PropTypes.string,
    })
  ),
  issuesRemaining: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      date: PropTypes.string,
    })
  ),
  tasksKept: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      instructions: PropTypes.string,
    })
  ),
  tasksAdded: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      instructions: PropTypes.string,
    })
  ),
};
