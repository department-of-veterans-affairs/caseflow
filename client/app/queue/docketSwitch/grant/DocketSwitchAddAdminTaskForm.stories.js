import React from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import { DocketSwitchAddAdminTaskForm } from './DocketSwitchAddAdminTaskForm';

export default {
  title: 'Queue/Docket Switch/DocketSwitchAddAdminTaskForm',
  component: DocketSwitchAddAdminTaskForm,
  decorators: [
    (storyFn) => {
      const methods = useForm();

      return <FormProvider {...methods}>{storyFn()}</FormProvider>;
    },
  ],
  parameters: {},
  args: {
    baseName: 'newTasks[0]',
    item: { type: null, instructions: '' },
  },
  argTypes: {
    onRemove: { action: 'onRemove' },
  },
};
const Template = (args) => <DocketSwitchAddAdminTaskForm {...args} />;

export const Basic = Template.bind({});
Basic.parameters = {
  docs: {
    storyDescription:
      'Used by attorney in Clerk of the Board office to complete a grant of a docket switch checkout flow ',
  },
};
