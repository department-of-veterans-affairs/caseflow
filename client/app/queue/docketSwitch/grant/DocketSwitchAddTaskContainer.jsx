import React, { useMemo } from 'react';
import { useSelector } from 'react-redux';
import { DocketSwitchAddTaskForm } from './DocketSwitchAddTaskForm';
import { appealWithDetailSelector, getAllTasksForAppeal } from '../../selectors';
import { useHistory, useParams } from 'react-router';

export const DocketSwitchAddTaskContainer = () => {
  const { appealId } = useParams();
  const { goBack } = useHistory();

  const appeal = useSelector((state) =>
    appealWithDetailSelector(state, { appealId })
  );

  const tasks = useSelector((state) =>
    getAllTasksForAppeal(state, { appealId })
  );

  const docketType = useSelector((state) =>
    state.docketSwitch.formData.docketType);

  const handleCancel = () => {
    // Add code to clear docketSwitch redux store
  };

  const handleBack = () => goBack();

  const switchableTasks = useMemo(() => {
    return tasks.filter((task) => task.canMoveOnDocketSwitch);
  }, [tasks]);

  const handleSubmit = () => {
    // Add stuff to redux store
  };

  return (
    <>
      <DocketSwitchAddTaskForm
        onCancel={handleCancel}
        onSubmit={handleSubmit}
        docketFrom={appeal.docketName}
        taskListing={switchableTasks}
        onBack={handleBack}
        docketTo={docketType}
      />
    </>
  );
};
