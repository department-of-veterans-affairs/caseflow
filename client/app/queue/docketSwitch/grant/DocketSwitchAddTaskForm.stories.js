import React from 'react';
import { DocketSwitchAddTaskForm } from './DocketSwitchAddTaskForm';

const tasks = [
  { taskId: 1, appealId: 1, type: 'TaskTypeA', label: 'Task Type A' },
  { taskId: 2, appealId: 1, type: 'TaskTypeB', label: 'Task Type B' },
  { taskId: 3, appealId: 1, type: 'TaskTypeC', label: 'Task Type C' },
];

export default {
  title: 'Queue/Docket Switch/DocketSwitchAddTaskForm',
  component: DocketSwitchAddTaskForm,
  decorators: [
    // AppSegment styling relies on being inside a .cf-content-inside element
    (storyFn) => <div className="cf-content-inside">{storyFn()}</div>,
  ],
  parameters: {},
  args: {
    appellantName: 'Jane Doe',
    taskListing: tasks,
    docketFrom: 'Direct Review',
    docketTo: 'Hearings',
  },
  argTypes: {
    onBack: { action: 'back' },
    onCancel: { action: 'cancel' },
    onSubmit: { action: 'submit' },
    type: {
      control: { type: 'select', options: ['Task Type A', 'Task Type B', 'Task Type C'] },
    },
  },
};
const Template = (args) => <DocketSwitchAddTaskForm {...args} />;
export const Basic = Template.bind({});
Basic.parameters = {
  docs: {
    storyDescription:
      'Used by attorney in Clerk of the Board office to complete a grant of a docket switch checkout flow ',
  },
};