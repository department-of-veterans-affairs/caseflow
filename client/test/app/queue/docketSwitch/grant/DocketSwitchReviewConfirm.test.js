import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { sub } from 'date-fns';
import { DocketSwitchReviewConfirm } from 'app/queue/docketSwitch/grant/DocketSwitchReviewConfirm';
import {
  DOCKET_SWITCH_GRANTED_CONFIRM_TITLE,
  DOCKET_SWITCH_GRANTED_CONFIRM_DESCRIPTION_B,
} from 'app/../COPY';

describe('DocketSwitchReviewConfirm', () => {
  const onSubmit = jest.fn();
  const onCancel = jest.fn();
  const onBack = jest.fn();
  const docketFrom = 'direct_review';
  const docketTo = 'hearing';
  const tasks = [
    {
      taskId: 1,
      appealId: 1,
      type: 'TaskTypeA',
      label: 'Task Type A',
      instructions: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    },
    {
      taskId: 2,
      appealId: 1,
      type: 'TaskTypeB',
      label: 'Task Type B',
      instructions: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    },
    {
      taskId: 3,
      appealId: 1,
      type: 'TaskTypeC',
      label: 'Task Type C',
      instructions: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    },
  ];

  const decisionDate = sub(new Date('2021-02-01'), { days: 10 });

  const issues = [
    {
      id: 1,
      program: 'compensation',
      description: 'PTSD denied',
      decision_date: decisionDate,
    },
    {
      id: 2,
      program: 'compensation',
      description: 'Left knee denied',
      decision_date: decisionDate,
    },
    {
      id: 3,
      program: 'compensation',
      description: 'Right knee denied',
      decision_date: decisionDate,
    },
    {
      id: 4,
      program: 'compensation',
      description: 'Right knee granted at 90%',
      decision_date: decisionDate,
    },
  ];

  const defaults = {
    docketFrom,
    docketTo,
    onBack,
    onCancel,
    onSubmit,
    originalReceiptDate: sub(new Date('2021-02-01'), { days: 30 }),
    docketSwitchReceiptDate: sub(new Date('2021-02-01'), { days: 7 }),
    issuesSwitched: issues,
    tasksKept: tasks,
    veteranName: 'Foo Bar',
  };

  const setup = (props) =>
    render(<DocketSwitchReviewConfirm {...defaults} {...props} />);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
    expect(
      screen.getByText(DOCKET_SWITCH_GRANTED_CONFIRM_TITLE)
    ).toBeInTheDocument();

    expect(
      screen.getByText(DOCKET_SWITCH_GRANTED_CONFIRM_DESCRIPTION_B)
    ).toBeInTheDocument();

    expect(
      screen.getByText(
        [
          'You are switching from Direct Review to Hearing.',
          'Tasks specific to the Direct Review docket will be automatically removed,',
          'and tasks associated with the Hearing docket will be automatically created.',
        ].join(' ')
      )
    ).toBeInTheDocument();
  });

  it('fires onCancel', async () => {
    setup();
    expect(onCancel).not.toHaveBeenCalled();

    await userEvent.click(screen.getByRole('button', { name: /cancel/i }));
    expect(onCancel).toHaveBeenCalled();
  });

  it('fires onBack', async () => {
    setup();
    expect(onBack).not.toHaveBeenCalled();

    await userEvent.click(screen.getByRole('button', { name: /back/i }));
    expect(onBack).toHaveBeenCalled();
  });

  it('fires onSubmit', async () => {
    setup();
    expect(onSubmit).not.toHaveBeenCalled();

    // submit and wait for async validation
    await act(async () =>
      userEvent.click(
        screen.getByRole('button', { name: /confirm docket switch/i })
      )
    );

    expect(onSubmit).toHaveBeenCalled();
  });

  describe('full grant', () => {
    it('renders issues correctly', async () => {
      setup();

      expect(
        screen.getByText('Issues switched to new docket')
      ).toBeInTheDocument();

      expect(
        screen.queryByText('Issues on original docket')
      ).not.toBeInTheDocument();
    });
  });

  describe('with alternate claimant', () => {
    it('renders the claimant row', async () => {
      const { container } = setup({ claimantName: 'Jane Doe' });

      expect(container).toMatchSnapshot();
    });
  });
});
