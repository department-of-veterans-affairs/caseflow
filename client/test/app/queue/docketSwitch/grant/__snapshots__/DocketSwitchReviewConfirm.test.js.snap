// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocketSwitchReviewConfirm renders correctly 1`] = `
<div>
  <section
    class="cf-app-segment cf-app-segment--alt"
  >
    <h1>
      Switch Docket: Review and Confirm
    </h1>
    <p>
      <strong>
        You are switching from Direct Review to Hearing. Tasks specific to the Direct Review docket will be automatically removed, and tasks associated with the Hearing docket will be automatically created.
      </strong>
    </p>
    <p>
      Please review all the information below before switching the docket. Once you confirm, tasks will be sent to the appropriate team or individual and the original docket number of the appeal will be preserved.
    </p>
    <table
      class="usa-table-borderless css-1t3jtyy"
    >
      <tbody>
        <tr>
          <td>
            Veteran
          </td>
          <td>
            Foo Bar
          </td>
        </tr>
        <tr>
          <td>
            VA Form - 10182 Receipt Date
          </td>
          <td>
            1/1/2021
          </td>
        </tr>
        <tr>
          <td>
            Docket Switch Receipt Date
          </td>
          <td>
            1/24/2021
          </td>
        </tr>
        <tr>
          <td>
            New Review option
          </td>
          <td>
            Hearing
          </td>
        </tr>
        <tr>
          <td>
            Issues switched to new docket
          </td>
          <td>
            <table
              class="usa-table-borderless"
            >
              <tbody>
                <tr>
                  <td>
                    <div>
                      1. PTSD denied
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      2. Left knee denied
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      3. Right knee denied
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      4. Right knee granted at 90%
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
        <tr>
          <td>
            Transferred/Added tasks
          </td>
          <td>
            <table
              class="usa-table-borderless"
            >
              <tbody>
                <tr>
                  <td>
                    <div>
                      <em />
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      1. Task Type A
                    </div>
                    <div>
                      <em>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </em>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      2. Task Type B
                    </div>
                    <div>
                      <em>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </em>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      3. Task Type C
                    </div>
                    <div>
                      <em>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </em>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
  <div
    class="controls cf-app-segment"
  >
    <div
      data-css-63oe3q=""
    >
      <span>
        <button
          class="cf-right-side usa-button"
          data-css-wn5k45=""
          id="button-next-button"
          type="button"
        >
          Confirm docket switch
        </button>
      </span>
      <button
        class="cf-btn-link usa-button"
        id="button-cancel-button"
        type="button"
      >
        Cancel
      </button>
      <button
        class="cf-right-side cf-prev-step usa-button-secondary usa-button"
        id="button-back-button"
        type="button"
      >
        Back
      </button>
    </div>
  </div>
</div>
`;

exports[`DocketSwitchReviewConfirm with alternate claimant renders the claimant row 1`] = `
<div>
  <section
    class="cf-app-segment cf-app-segment--alt"
  >
    <h1>
      Switch Docket: Review and Confirm
    </h1>
    <p>
      <strong>
        You are switching from Direct Review to Hearing. Tasks specific to the Direct Review docket will be automatically removed, and tasks associated with the Hearing docket will be automatically created.
      </strong>
    </p>
    <p>
      Please review all the information below before switching the docket. Once you confirm, tasks will be sent to the appropriate team or individual and the original docket number of the appeal will be preserved.
    </p>
    <table
      class="usa-table-borderless css-1t3jtyy"
    >
      <tbody>
        <tr>
          <td>
            Veteran
          </td>
          <td>
            Foo Bar
          </td>
        </tr>
        <tr>
          <td>
            Claimant
          </td>
          <td>
            Jane Doe
          </td>
        </tr>
        <tr>
          <td>
            VA Form - 10182 Receipt Date
          </td>
          <td>
            1/1/2021
          </td>
        </tr>
        <tr>
          <td>
            Docket Switch Receipt Date
          </td>
          <td>
            1/24/2021
          </td>
        </tr>
        <tr>
          <td>
            New Review option
          </td>
          <td>
            Hearing
          </td>
        </tr>
        <tr>
          <td>
            Issues switched to new docket
          </td>
          <td>
            <table
              class="usa-table-borderless"
            >
              <tbody>
                <tr>
                  <td>
                    <div>
                      1. PTSD denied
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      2. Left knee denied
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      3. Right knee denied
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      4. Right knee granted at 90%
                    </div>
                    <div>
                      Decision date:
                       
                      1/21/2021
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
        <tr>
          <td>
            Transferred/Added tasks
          </td>
          <td>
            <table
              class="usa-table-borderless"
            >
              <tbody>
                <tr>
                  <td>
                    <div>
                      <em />
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      1. Task Type A
                    </div>
                    <div>
                      <em>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </em>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      2. Task Type B
                    </div>
                    <div>
                      <em>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </em>
                    </div>
                  </td>
                </tr>
                <tr>
                  <td>
                    <div>
                      3. Task Type C
                    </div>
                    <div>
                      <em>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit
                      </em>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
  <div
    class="controls cf-app-segment"
  >
    <div
      data-css-63oe3q=""
    >
      <span>
        <button
          class="cf-right-side usa-button"
          data-css-wn5k45=""
          id="button-next-button"
          type="button"
        >
          Confirm docket switch
        </button>
      </span>
      <button
        class="cf-btn-link usa-button"
        id="button-cancel-button"
        type="button"
      >
        Cancel
      </button>
      <button
        class="cf-right-side cf-prev-step usa-button-secondary usa-button"
        id="button-back-button"
        type="button"
      >
        Back
      </button>
    </div>
  </div>
</div>
`;
